### PR TITLE
added options to config, make and install

### PR DESCRIPTION
**Pull request by Juppit**

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -428,49 +441,50 @@ $(XTDLP)/$(XTENSA_TOOLCHAIN_LINUX_TAR):
...
 
 
 
-libhal: $(TOOLCHAIN)/xtensa-lx106-elf/lib/libhal.a
+build_libhal: $(TOOLCHAIN)/xtensa-lx106-elf/lib/libhal.a
 
-$(TOOLCHAIN)/xtensa-lx106-elf/lib/libhal.a: toolchain $(XTDLP)/$(LIBHAL_DIR)/configure.ac
-	make -C $(XTDLP)/$(LIBHAL_DIR) -f ../../Makefile _libhal
+#$(TOOLCHAIN)/xtensa-lx106-elf/lib/libhal.a: toolchain $(XTDLP)/$(LIBHAL_DIR)/configure.ac
+$(TOOLCHAIN)/xtensa-lx106-elf/lib/libhal.a: $(XTDLP)/$(LIBHAL_DIR)/configure.ac
```
Comment (DimmKirr) [inline Makefile:480]: What if there is no toolchain built, how will we build libhal? why do we remove `toolchain` dependency?

Comment (Juppit) [inline Makefile:480]: Hi,
I like the idea, that a make runs once, and in a second run there is 
nothing to do.
Therefore I tried different things, one is to touch the build.

For libhal a tried this and that, but no really success.

Ok, at least the following idea:

I did a more orthogonal rule with a build directory as tommie did it.
I guess separating source and build is better.
And then I adapted it to the other targets in your Makefile:

> # Libhal
> 
> $(XTDLP)/$(LIBHAL_DIR)/build: $(XTDLP)/$(LIBHAL_DIR)/configure.ac
>   cd $(XTDLP)/$(LIBHAL_DIR); autoreconf -i
>   mkdir -p $(XTDLP)/$(LIBHAL_DIR)/build/
>   cd $(XTDLP)/$(LIBHAL_DIR)/build/; PATH=$(SAFEPATH) ../$(CONF_OPT) --host=$(TARGET) --prefix=$(TOOLCHAIN)/xtensa-lx106-elf/
>   PATH=$(SAFEPATH) $(MAKE_OPT) -C $(XTDLP)/$(LIBHAL_DIR)/build/
>   @touch $@
> 
> $(XTDLP)/$(LIBHAL_DIR): $(XTDLP)/$(LIBHAL_DIR)/build
>   PATH=$(SAFEPATH) make $(INST_OPT) -C $(XTDLP)/$(LIBHAL_DIR)/build/
>   @touch $@

I added build-libhal

> build-libhal: get-src build-binutils $(XTDLP)/$(LIBHAL_DIR)/build $(XTDLP)/$(LIBHAL_DIR)

and added libhal to toolchain

> $(TOOLCHAIN)/bin/xtensa-lx106-elf-gcc: $(TOOLCHAIN) $(XTDLP) ... build-second-stage-gcc build-libhal strip compress-upx

For convenience I added this test (as temporary hack)

> GCC_DIR = gcc-$(GCC_VERSION)
> 
> ifneq "$(wildcard $(XTDLP)/gcc-xtensa)" ""
>     GCC_DIR := gcc-xtensa
> endif

so we have not to care for this difference between rtos-sdk, master, 
develop.

Please tell me your opinion about these changes, so I can make a PR.

On 08.01.2016 at 04:43 wrote Dmitry Kireev:

File: Makefile [hunk 2]
```diff
@@ -428,49 +441,50 @@ $(XTDLP)/$(XTENSA_TOOLCHAIN_LINUX_TAR):
...
 
 $(XTDLP)/$(ESPTOOL2_DIR)/esptool2.c:
 	@echo "You cloned without --recursive, fetching esptool2 for you."
-	git submodule update --init src/$(ESPTOOL2_DIR)
+	git submodule update --init $(XTDLP)/$(ESPTOOL2_DIR)
 
 $(XTDLP)/$(MEMANALYZER_DIR)/MemAnalyzer.sln:
 	@echo "You cloned without --recursive, fetching MemAnalyzer for you."
-	git submodule update --init src/$(MEMANALYZER_DIR)
+	git submodule update --init $(XTDLP)/$(MEMANALYZER_DIR)
```
Comment (DimmKirr) [inline Makefile:473]: Have you actually tested `git submodule update --init $(XTDLP)/$(MEMANALYZER_DIR)` ? Absolute path (which is $(XTDLP)) didn't work for me.

File: Makefile (hunk 1 above)
Comment (DimmKirr) [inline Makefile:480]: > $(TOOLCHAIN)/bin/xtensa-lx106-elf-gcc: $(TOOLCHAIN) $(XTDLP) ... build-second-stage-gcc build-libhal strip compress-upx

where? I see

> $(TOOLCHAIN)/bin/xtensa-lx106-elf-gcc: $(TOOLCHAIN) $(XTDLP) $(XTBP) build-gmp build-mpfr build-mpc build-binutils build-gdb build-first-stage-gcc build-newlib build-second-stage-gcc strip compress-upx 

currently

Comment (DimmKirr) [inline Makefile:480]: Also, to my understanting, libhal is built using xtensa target, so why would you depend on libhal in building actual xtensa compiler?

File: Makefile (hunk 2 above)
Comment (Juppit) [inline Makefile:473]: I tested it for esp-binutils, gcc, gdb, gmp, libhal, mpc ..
not for utils.
When it's not working for you (on mingw?) we should change it to src, no problem

File: Makefile (hunk 1 above)
Comment (Juppit) [inline Makefile:480]: I'm having trouble with these dependencies...
But I see the make will build libhal again and again. I mean after the first make, a second one builds it again, but I see no reason.
That's why I made the new proposal in my comment (not in this PR), which makes things clearer, I hope..

**Pull request merged**